docs/source/json-api/index.rst

```diff
@@ -540,10 +540,7 @@ HTTP Request
 
 Where:
 
-- ``templateId`` is the contract template identifier, which can be formatted as either:
-
-  + ``"<package ID>:<module>:<entity>"`` or
-  + ``"<module>:<entity>"`` if contract template can be uniquely identified by its module and entity name.
+- ``templateId`` is the contract template identifier, which is formatted as ``"<package ID>:<module>:<entity>"``. As a convenience for rapid prototyping, you can also omit the package ID (i.e. specifying the ``templateId`` as ``"<module>:<entity>"``), but you are recommended to explicitly specifying the package ID as soon as you move away from the prototyping phase of development, since it's common to have more versions of a template sharing the same module and entity name but with different package IDs. If the package identifier is not specified and the template cannot be uniquely identified without it, the HTTP JSON API service will report that the specified template cannot be found.
```

Comment:
```suggestion
- ``templateId`` is the contract template identifier, which is formatted as ``"<package ID>:<module>:<entity>"``. As a convenience for rapid prototyping, you can also omit the package ID (i.e. specifying the ``templateId`` as ``"<module>:<entity>"``) **if there is only one template with that name across _all loaded packages_**. We strongly recommend you explicitly specify the package ID as soon as you move away from the prototyping phase of development, since it's common to have more versions of a template sharing the same module and entity name but with different package IDs. If the package identifier is not specified and the template cannot be uniquely identified without it, the HTTP JSON API service will report that the specified template cannot be found.
```

Because more explicit scary messaging is always better.

Reply:
We could go further, and explicitly state something like:
```suggestion
- ``templateId`` is the contract template identifier, which is formatted as ``"<package ID>:<module>:<entity>"``. As a convenience for interactive API exploration (such as with curl and similar tools), you can also omit the package ID (i.e. specifying the ``templateId`` as ``"<module>:<entity>"``) **if there is only one template with that name across _all loaded packages_**. Code should always specify the package ID, since it's common to have more versions of a template sharing the same module and entity name but with different package IDs. If the package identifier is not specified and the template cannot be uniquely identified without it, the HTTP JSON API service will report that the specified template cannot be found. **Omitting the package ID is not supported for production use.**
```

Reply:
I like this version, it will prevent a lot more problems I think.

Reply:
When I accepted the suggestion the Github UI didn't show the second suggestion. Addressed in https://github.com/digital-asset/daml/pull/14601/commits/93ace8a8574b6c11a540809923e94cbf7f27039f.